src/Shared/Diagnostics/ActivityCreator.cs

```diff
@@ -0,0 +1,106 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace Microsoft.AspNetCore.Shared;
+
+internal static class ActivityCreator
+{
+    /// <summary>
+    /// Create an activity with details received from a remote source.
+    /// </summary>
+    public static Activity? CreateFromRemote(
+        ActivitySource activitySource,
+        DistributedContextPropagator propagator,
+        object distributedContextCarrier,
+        DistributedContextPropagator.PropagatorGetterCallback propagatorGetter,
+        string activityName,
+        IEnumerable<KeyValuePair<string, object?>>? tags,
+        IEnumerable<ActivityLink>? links,
+        bool diagnosticsOrLoggingEnabled)
+    {
+        propagator.ExtractTraceIdAndState(
+            distributedContextCarrier,
+            propagatorGetter,
+            out var requestId,
+            out var traceState);
+
+        Activity? activity = null;
+        if (activitySource.HasListeners())
+        {
+            if (ActivityContext.TryParse(requestId, traceState, isRemote: true, out ActivityContext context))
+            {
+                // The requestId used the W3C ID format. Unfortunately, the ActivitySource.CreateActivity overload that
+                // takes a string parentId never sets HasRemoteParent to true. We work around that by calling the
+                // ActivityContext overload instead which sets HasRemoteParent to parentContext.IsRemote.
+                // https://github.com/dotnet/aspnetcore/pull/41568#discussion_r868733305
+                activity = activitySource.CreateActivity(activityName, ActivityKind.Server, context, tags: tags, links: links);
+            }
+            else
+            {
+                // Pass in the ID we got from the headers if there was one.
+                activity = activitySource.CreateActivity(activityName, ActivityKind.Server, string.IsNullOrEmpty(requestId) ? null! : requestId, tags: tags, links: links);
```

Comment:
>string.IsNullOrEmpty(requestId) [](http://example.com/codeflow?start=92&length=31)

Q: why you need this check if you pass null anyway? is it expected to get empty strings at all?

Reply:
I don't know. This is code I moved. I left it as it is because I want to minimize unrelated changes.